lib/puppet/agent.rb

```diff
@@ -99,4 +116,25 @@ def with_client(transaction_uuid)
   ensure
     @client = nil
   end
+
+  # Handle agent timeouts
+  #
+  # This method handles timeouts raised when the Puppet agent takes too
+  # long to run. An error message is logged, the Agent application is set
+  # to a stopping state and an exit code of 1 returned.
+  #
+  # @param exception [Puppet::Agent::RunTimeoutError] The exception raised by
+  #   `Timeout.timeout` when an agent run exceeds the limit set by the
+  #   `:runtimeout` setting.
+  #
+  # @return [Integer] The exit code for the agent process. Set to 1.
+  def handle_runtimeout(exception)
+    Puppet.log_exception(exception, _("Execution of %{client_class} did not complete within %{runtimeout} seconds and was terminated.") %
+      {client_class: client_class,
+      runtimeout: Puppet[:runtimeout]})
+
+    Puppet::Application.stop!
```

Comment:
is there a reason to call this outside of the client run?

Reply:
I recommended that Charlie look at the kinds of things we do in our signal handlers to see if there was application clean up that we can/should do after throwing the exception.

It looks like calling `Puppet::Application.stop!` is really just setting a flag so that other code can see that Puppet wants to stop and act accordingly (like cancel a transaction).

Is it safe to assume that at this point we've interrupted any code that would check this flag and we can go ahead and exit similarly to what we do in the other rescue blocks?

Reply:
If that's the case and we remove the call to `Application.stop!` then I don't think there's any harm in also moving the contents of this method up to the rescue block (it then wouldn't be any more complicated than the other blocks.

Reply:
Setting an expectation on a method call is much easier to test than a rescue block though.

Reply:
You have an opinion here, @pcarlisle ?

Reply:
Okay, I think it's safe to assume that there's nothing that will respect `Application.stop!` after we've interrupted the run at this level with the timeout - so we don't need to call it here and should remove it.

I also think if we're not really doing any special handling then we don't need this additional method. You mentioned testing on a method invocation being easier than a rescue block, you can test the behavior by testing the lack of a method call though. ie (in pseudo code)
```rb
      it 'times out when a run exceeds the set limit' do
         client = AgentTestClient.new
         client.instance_eval do

          def processing; end
          def handler; end

          def run(client_options = {})
            begin
              ::Kernel.sleep(5)
              processing()
            rescue
              handling()
            end
          end
        end

        AgentTestClient.expects(:new).returns client
        expect(@agent).not_to receive(:processing)
        expect(@agent).not_to receive(:handling)

        expect(@agent.run).to eq(1)
      end

```

Reply:
Sounds good. Pushed up an update with the proposed changes.